compiler/parser/RdrHsSyn.hs

```diff
@@ -3092,75 +3112,74 @@ mkTemplateDecls templateName fields decls = do
       choiceInstanceDecls = concatMap (mkChoiceInstanceDecl templateName) choicesWithArchive
       choiceDecls = concatMap (mkChoiceDecls (getLoc templateName) conName sharedBinds) choicesWithArchive
       keyInstanceDecl = mkKeyInstanceDecl sharedBinds templateName conName vt
-      templateInterfaceImplements = map (mkInterfaceInstance vtTemplateName conName sharedBinds) vtImplements
-      templateImplementsMarkers = concatMap (mkImplementsMarker vtTemplateName) vtImplements
+      adjustImplementsDeclBlock implements =
```

Comment:
there's probably a better name for this?

Reply:
is there a reason why we need to do this posthoc adjustment instead of doing it directly in `mkInterfaceImplements` thereby potentially also eliminating the question of what we call it?

Reply:
there is! I also call `mkInterfaceImplements` from `mkInterfaceDecl` to construct the methods for the interface itself (`_method_Token_Token_<method>`), and those don't need the lambda abstraction for `this`, since they're just defined as the top level method.

Example:

```Haskell
instance DA.Internal.Desugar.HasMethod Token "noopImpl" (() -> Update ())

noopImpl : DA.Internal.Desugar.Implements t Token => t -> () -> Update ()
noopImpl t = GHC.Types.primitiveInterface @"noopImpl" (DA.Internal.Desugar.toInterface @_ @Token t)

-- Needs to bind `this` in a lambda abstraction
_method_Asset_Token_noopImpl = 
  DA.Internal.Desugar.mkMethod @Asset @Token @"noopImpl" (\ this@Asset {..} -> let _ = this in \ nothing -> do pure ())

-- Doesn't need to bind `this` since `noopImpl` already has a `t` argument.
_method_Token_Token_noopImpl = 
  DA.Internal.Desugar.mkMethod @Token @Token @"noopImpl" (noopImpl)
```

But it probably still makes sense to factor this out somehow? I could add an `LHsExpr GhcPs -> LHsExpr GhcPs` argument to `mkInterfaceDecl`, calling it with `bindThis` in the former case and with `id` in the latter.

Reply:
Thanks for the explanation.

I feel like part of what makes this slightly confusing is that we’re conflating a few different things:
1. The expression the user defined in their `implements` block represented via `ImplementsFunction`.
2. The expression we use in the `_method_`.

Maybe using different types for both and then having this adjust thingy move from one to the other would make things clearer?

Reply:
As you mentioned on slack, the definitions for `_method_Token_Token_<method>` aren't really used, so I can just move the lambda abstraction into `mkInterfaceDecl`. This way, the distinction between 1. and 2. is internal to `mkInterfaceDecl`